File: java/yass/test/ch/softappeal/yass/remote/test/ContractIdTest.java
package ch.softappeal.yass.remote.test;

import ch.softappeal.yass.remote.ContractId;
import ch.softappeal.yass.remote.TaggedMethodMapper;
import ch.softappeal.yass.test.InvokeTest;
import org.junit.Assert;
import org.junit.Test;

public class ContractIdTest {

    public static final ContractId<InvokeTest.TestService> ID = ContractId.create(InvokeTest.TestService.class, 987654, TaggedMethodMapper.FACTORY);

    @Test public void test() {
        Assert.assertSame(InvokeTest.TestService.class, ID.contract);
        Assert.assertEquals(987654, ID.id);
    }

}

File: kotlin/yass/test/ch/softappeal/yass/remote/ContractIdTest.kt
package ch.softappeal.yass.remote

import ch.softappeal.yass.*
import kotlin.test.*

private abstract class Role : Services(SimpleMethodMapperFactory)

class ContractIdTest {
    @Test
    fun contractId() {
        val contractId = contractId<Calculator>(123, SimpleMethodMapperFactory)
        assertEquals(123, contractId.id)
        assertSame(Calculator::class.java, contractId.contract)
    }

    @Test
    fun javaContractId() {
        val contractId = contractId(JavaCalculator::class.java, 123, SimpleMethodMapperFactory)
        assertEquals(123, contractId.id)
        assertSame(JavaCalculator::class.java, contractId.contract)
    }

    @Test
    fun okServices() {
        class Initiator : Role() {
            val calculator0 = contractId<Calculator>(0)
            val calculator1 = contractId<Calculator>(1)
        }

        val initiator = Initiator()
        assertEquals(0, initiator.calculator0.id)
        assertSame(Calculator::class.java, initiator.calculator0.contract)
        assertEquals(1, initiator.calculator1.id)
        assertSame(Calculator::class.java, initiator.calculator1.contract)
    }

    @Test
    fun duplicatedServices() {
        class Initiator : Role() {
            val calculator0 = contractId<Calculator>(0)
            val calculator1 = contractId<Calculator>(0)
        }
        assertEquals(
            "service with id 0 already added",
            assertFailsWith<IllegalArgumentException> { Initiator() }.message
        )
    }
}
